Testament: test process batching and internals rework

saem opened the pull request:
Started as introducing knownIssue and disabled, but reworking internals heavily and minor consequences to the interface.

Refactor Consequences:
- introduces a more data oriented design
- simplifies categories
  - remove dead/unnecessary ones (IC/Navigator)
  - remove the megatest category hack
  - likely other changes before finishing
- makes js a first class target
  - now a default execution target
  - it should be explicitly excluded (likely need a feature)

break changes forecast:
- this will necessitate some test/spec clean-up
- might run a fair bit slower without megatest
- although in some areas it's more parallelizeable (always runs in parallel)
- removed ic and navigator categories, need to fix spec parsing 

reviewable things:
- ExecutionState is getting pretty stable
- The batching is somewhat final, but making a bit lazier can avoid IO
  - can't quite add that to the stack of things to change yet
- all, category, and single test all behave the same now
- should we bother with pattern?

Known issues:
- need to get an end to end execution going
- runTests is a beast
- lol docs
- tests for testament

Lessons learned:
Many, biggest one so far is orchestration of dependencies and testing
are separate things and the existing testament conflates these.

<!--
## Finalizing a PR:

### title:
* reads like a short changelog line

### body:
* describe the current behaviour
* describe why this particular approach
* if it's breaking, migration steps
* note any follow-on work

### content:
* leave code better than you found it, add docs, tests, etc
-->

<!--
* note any issues that this fixes entirely
* use `fixes #123 .` to auto-close issues
-->

saem committed:
formatting - testament/specs.parseTargets

saem committed:
testament - formatting & remove dead code

saem committed:
sketch knownIssue spec

saem committed:
partially working known issue retesting

saem committed:
partial refactor of testament internals

- introduces a more data oriented design
- simplifies categories
- makes js a first class target

break changes forecasting:
- this will necessitate some test/spec clean-up
- might run a fair bit slower without megatest
- although in some areas it's more parallelizeable
- removed ic and navigator categories, need to fix spec parsing first

reviewable things:
- ExecutionState is getting pretty stable
- The batching is somewhat final, but making a bit lazier can avoid IO
  - can't quite add that to the stack of things to change yet
- all, category, and single test all behave the same now
- should we bother with pattern?

Known issues:
- need to get an end to end execution going
- runTests is a beast
- tests for testament

Lessons learned:
Many, biggest one so far is orchestration of dependencies and testing
are separate things and the existing testament conflates these.

saem commented on testament/testament.nim:
```diff
@@ -851,4 +884,437 @@ proc main() =
+    of actionCompile:
+      partOfRun: bool
+
+  Execution = object
+    # user and execution inputs
+    filter: TestFilter       ## filter that was configured
+    flags: ExecutionFlags    ## various options set by the user
+    targets: TestTargets     ## specified targets or `noTargetsSpecified`
+    workingDir: string       ## working directory to begin execution in
+    nimSpecified: bool       ## whether the user specified the nim
+    testArgs: string         ## arguments passed to tests by the user
+
+    # environment input / setup
+    compilerPath: string     ## compiler command to use
+    testsDir: string         ## where to look for tests
+
+    # test discovery data
+    testCats:  Categories    ## categories discovered, for this execution
+    testFiles: seq[TestFile] ## files for this execution
+    testSpecs: seq[TSpec]    ## spec for each file
+
+    # test execution data
+    testRuns: seq[TestRun]   ## a test run: reject, compile, or compile + run
+    actions: seq[TestAction] ## test actions for each run, phases of a run
+    debugInfo: DebugInfo     ## debug info related to actions run for tests
+
+    # test execution related data
+    retryList: RetryList     ## list of failures to potentially retry later
+
+  ParseCliResult = enum
+    parseSuccess       ## successfully parsed cli params
+    parseQuitWithUsage ## parsing failed, quit with usage message
+
+const
+  testResultsDir = "testresults"
+  cacheResultsDir = testResultsDir / "cacheresults"
+  noTargetsSpecified: TestTargets = {}
+  defaultExecFlags = {outputColour}
+  defaultBatchSize = 10
+  noMatrixEntry: EntryId = -1
```
@alaviss as you can see I learned similar lessons with DOD. 😆

alaviss replied:
You really should turn those `int` into `int32`, no one do more than 2M tests yet

saem replied:
Done locally.

saem committed:
removed examples category

Also outputs the tests it'll run, this demonstrated issues with tests
that specify cmd, such as arc tests.

saem committed:
removed niminaction tests

- they're dead code, most of them are simply compiles
- many are covered far more comprehensively in the spec
- the async ones are gone with async removal
- the error message related ones will change dramatically
- threading is so broken I don't see how these help

saem committed:
testament: remove nim in action remenant

saem committed:
got test compilation working

next major todos:
- execution
- skips/known issues
- reporting

saem committed:
added special handling for gc tests

need to add in the test matrix override handling

miles to go.

saem committed:
just need to rough in stdlib

then can work through creating test batches

saem committed:
compiling all the tests

next up is reporting and all that.

saem committed:
part way through wiring up results

saem committed:
further along creating the run command

saem committed:
started executing compiled executables

saem committed:
set reminders before rebase

saem committed:
test file name handling is awful, still

saem committed:
lift out the ridiculous template

saem committed:
part way through error reporting

- starting to emit program run success/errors
- need to handle compile and reject

saem committed:
right before I try another big refactor

saem committed:
end to end runs starting to work

next up is bugs and lots of code clean-up

saem committed:
remove some cruft

immediate next step is to handle skips and known issues

saem committed:
remove io test special category

the io category had one test that was disabled long ago

removed it and older helper code `readall_echo.nim`

saem committed:
add skip file handling

saem committed:
remove commented out code

saem committed:
setup simulate/dry run

saem committed:
skipped tests are reported

skipping is either because they're disabled or known issues

there seems to be an issue with tests discovery of execution,
not all tests seem to be run.

saem committed:
lang spec: don't test union pragma in js

saem committed:
separate test result output reporting

saem committed:
skip run actions if compile action fails

saem committed:
lang spec: don't test ptr boundchecks pragma in js

saem committed:
fix checking of next batch size

check if the next batch is full after failed action pruning

saem committed:
fix tests so they work on more targets

saem committed:
mark t03_inheritance with a known issue

but it only applies to js... hmm

saem committed:
fix tests so they pass on js

saem committed:
codecheck and nimcache dir fixes

saem committed:
fix ambsym test so it works on the js target

saem committed:
arc test clean-up; 3 failures

failure info:
- arc/tcontrolflow fails in cpp, not sure why
- arc/thavlak_orc_stress fails due to valgrind issues in c and cpp
- arc/torc_selfcycles fails similar to thavlak_orc_stress in c and cpp

saem committed:
have categories specify test target defaults

point here is that category commonalities should be in category handling

haxscramper commented:
@saem aside from `xxx` todos, maybe failing tests and rebase, is there anything else that needs to be done for this PR? I want to continue this after https://github.com/nim-works/nimskull/pull/476 is merged and initial test directory cleanup is finished.

haxscramper commented:
Many changes from this PR are already implemented so this can be closed and PRed in a simplified manner to avoid dealing with all merge conflicts.
